### ISSUE TITLE
model.save()/ load_model() inconsistency

### ISSUE THREAD

**Issue by ErikEngerd** [keras-team/keras#14181]
Environment: centos 7 
Python version: 3.7.6 (anaconda)
Tensorflow: 2.3.0

The script below creates a very simple model for the CIFAR dataset, does 1 epoch of training and prints the test accuracy. Next, it saves the model and deletes it from memory. Then it loads it again, and again evaluates the test accuracy. 

Expected result: test accuracy is the same. 
Actual result: test accuracy is different (in particular zero).

The issue occurs with  and without GPU (CUDA_VISIBLE_DEVICES=-1). 

```import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Conv2D, MaxPooling2D

print("Tensorflow version" + tf.__version__)


#####################
# load data

(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()
x_train = x_train / 255.0
x_test = x_test / 255.0

# use less data to speed up
x_train = x_train[:100]
y_train = y_train[:100]
x_test = x_test[:100]
y_test = y_test[:100]

#####################
# utility function

def get_test_accuracy(model, x_test, y_test):
    test_loss, test_acc = model.evaluate(x=x_test, y=y_test, verbose=0)
    print('=== accuracy: {acc:0.3f}'.format(acc=test_acc))
    
#####################
# define model

model = Sequential([
    Flatten(name='flatten'),
        Dense(units=10, activation='softmax')
])
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs = 1, verbose=0)
print("=== Test accuracy after training and before saving")
get_test_accuracy(model, x_test, y_test)

####################
# save model

modelpath="mymodel"
model.save(modelpath)
del model

####################
# load model

model = tf.keras.models.load_model(modelpath)
print("=== Test accuracy after loading")
get_test_accuracy(model, x_test, y_test)
```

I get the following output when running without GPU (the lines starting with === are the relevant lines): 

```
(base) erik@buzzard:~/notebooks/saveandload> CUDA_VISIBLE_DEVICES=-1 python saveloadproblem.py 
2020-08-01 20:49:35.806453: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
Tensorflow version2.3.0
2020-08-01 20:49:37.410489: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcuda.so.1
2020-08-01 20:49:37.723113: E tensorflow/stream_executor/cuda/cuda_driver.cc:314] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2020-08-01 20:49:37.723165: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:169] retrieving CUDA diagnostic information for host: buzzard
2020-08-01 20:49:37.723181: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:176] hostname: buzzard
2020-08-01 20:49:37.723273: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:200] libcuda reported version is: 450.51.6
2020-08-01 20:49:37.723333: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:204] kernel reported version is: 450.51.6
2020-08-01 20:49:37.723348: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:310] kernel version seems to match DSO: 450.51.6
2020-08-01 20:49:37.723649: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-08-01 20:49:37.730960: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 2208000000 Hz
2020-08-01 20:49:37.732509: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x5599eb89eb80 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-08-01 20:49:37.732530: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
=== Test accuracy after training and before saving
=== accuracy: 0.070
WARNING:tensorflow:From /home/erik/anaconda3/lib/python3.7/site-packages/tensorflow/python/training/tracking/tracking.py:111: Model.state_updates (from tensorflow.python.keras.engine.training) is deprecated and will be removed in a future version.
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.
2020-08-01 20:49:38.120876: W tensorflow/python/util/util.cc:348] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
WARNING:tensorflow:From /home/erik/anaconda3/lib/python3.7/site-packages/tensorflow/python/training/tracking/tracking.py:111: Layer.updates (from tensorflow.python.keras.engine.base_layer) is deprecated and will be removed in a future version.
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.
=== Test accuracy after loading
=== accuracy: 0.000
```

- [x ] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps

- [x ] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [ x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [ x

] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by ErikEngerd** [comment]
I have just checked that by downgrading the tensorflow-gpu package to version 2.2.0 (pip install tensorflow-gpu==2.2.0), it works.  Looks like an issue might have been introduced in tensorflow 2.3.0.

**Comment by ErikEngerd** [comment]
The issue is also reproducible on google colab with tensorflow 2.3.0. It does not occur with 2.2.0. By the way, the accuracy reported after loading the model is not 0 but another value which is still different from that before the model was saved. 

See [here](https://drive.google.com/file/d/1jbx4u0sk-7a3_AjXHj5ppe_0AxkYTwyP/view?usp=sharing)

**Comment by Chinmay922** [comment]
I am looking into this issue.

**Comment by Chinmay922** [comment]
Windows - 10
Cloud GPU - Google Colab
Python - 3.8.5
I tried working on your code provided and was able to save an load the model without any issues in TensorFlow 2.3.0 and the accuracy was not 0. Also, I gave a custom data set to train the model. Displayed the number of images in the dataset for Training and Test separate.  There are a few changes in the code and after running the code with a different data set I got a positive result for both the versions of Tensorflow.

Here is my version of the code for your reference,

# -*- coding: utf-8 -*-
"""cifar_dataset.ipynb

Automatically generated by Colaboratory.

Original file is located at
    https://colab.research.google.com/drive/1E9Rf6ZpzMKQvN1Zm7srpVxrSbVg2DghA
"""

import tensorflow as tf

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Conv2D, MaxPooling2D

print("Tensorflow version" + tf.__version__)

#from google.colab import drive
#drive.mount('/content/gdrive/',force_remount=True)
#root_path='/content/gdrive/My Drive/Colab Notebooks'
#dir_data=root_path 
#print(dir_data)
#dir_seg=dir_data+"/annotations_prepped_train"
#dir_img=dir_data+"/images_prepped_train"
#print("############################")

from keras.datasets import cifar10

""" import necessary libraries"""
import cv2,os
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import pyplot

# load dataset
(trainX, trainy), (testX, testy) = cifar10.load_data()
# Displaying the number of images in this dataset for Training and Test separate
print('Training_Shape: X=%s, y=%s' % (trainX.shape, trainy.shape))
print('Test_Shape: X=%s, y=%s' % (testX.shape, testy.shape))

""" Displaying the images in the data set """

for i in range(9):
  pyplot.subplot(330 + 1 + i)
  pyplot.imshow(trainX[i])

# Display the figure with subplots
pyplot.show()

#trainX=trainX/255.0
#testX=testX/255.0
pyplot.imshow(trainX[5])
pyplot.show()

def get_test_accuracy(model, testX, testy):
    test_loss, test_acc = model.evaluate(x=testX, y=testy, verbose=0)
    print('=== accuracy: {acc:0.3f}'.format(acc=test_acc))

model = Sequential([
    Flatten(name='flatten'),
        Dense(units=10, activation='softmax')
])
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(trainX, trainy, epochs = 1, verbose=0)
print("=== Test accuracy after training and before saving")
get_test_accuracy(model, testX, testy)



# save model

modelpath="mymodel"
model.save(modelpath)
del model

# load model

model = tf.keras.models.load_model(modelpath)
print("=== Test accuracy after loading")
get_test_accuracy(model, testX, testy)

**Comment by iteachmachines** [comment]
There might be some issue with tensorflow 2.3.0.It should work with tensorflow 2.1.0.

**Comment by ErikEngerd** [comment]
I tested also with tensorflow 2.2 0 and that works. Did not test with 2.1.0.

On 4 August 2020 07:11:06 CEST, priyam kakati <notifications@github.com> wrote:
>There might be some issue with tensorflow 2.3.0.It should work with
>tensorflow 2.1.0.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/keras-team/keras/issues/14181#issuecomment-668382947

Erik Brakkee

**Comment by iteachmachines** [comment]
@ErikEngerd  Was working fine with tensorflow 2.1.0.I've tested it.

**Comment by ErikEngerd** [comment]
can you confirm that it is *not* working with tensorflow 2.3.0 then? 

On 4 August 2020 23:09:55 CEST, priyam kakati <notifications@github.com> wrote:
>@ErikEngerd  Was working fine with tensorflow 2.1.0.I've tested it.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/keras-team/keras/issues/14181#issuecomment-668826638

Erik Brakkee

**Comment by kraiskil** [comment]
Bumped into the same issue.
Ubuntu 18.04, no GPU available.
Tensorflow installed with pip, python 3.6.

Downgrading 2.3.0 -> 2.2.0 made the save/load cycle work as expected.

**Comment by sushreebarsa** [comment]
@ErikEngerd Moving this issue to closed status as there has been no recent activity.In case you still face the error please create a new issue,we will get you the right help.Thanks!

**Comment by kraiskil** [comment]
@sushreebarsa was there something done in respect to this that you think it might work now?
